ChartTabs: Optimize performance using the data layer

jsnmoon opened the pull request:
This focuses on improving performance and maintainability of the Stats Dashboard (`/stats/day`, `/stats/week`, `/stats/month`, `/stats/year`). Namely, it achieves the following:

- Reducing redundancy in the two API requests used to populate the data.
  - Previously, we made a request for the current chart tab and another request for all tabs.
  - Now, we make a request for the current chart tab and another for the unselected tabs.
- Implements a Redux state using Calypso's data layer (+ schemas and tests).
- Reduces the amount of computation done on every render.

#### Testing instructions

1. Check out this branch locally and run the client tests; ensure they pass.
```shell
npm run test-client
```
2. Spin this branch up locally or via calypso.live.
3. Navigate to [`/stats/day`](http://calypso.localhost:3000/stats/day) and select a site. 
4. Ensure that the graph loads as expected. Also, verify that only two API requests to `/stats/visits` are made in the network console in your browser.

jmdodd commented:
@jsnmoon How did you measure and compare performance with parallelized network requests versus with the original two API `stats/visits` requests?

jsnmoon commented:
@jmdodd: Following up from our discussion on Slack, I've removed the parallelized network requests from this PR and have restored the original request strategy. 

I've also introduced an optimization to the original request strategy to reduce redundancy. Let's say the user has the `Views` tab selected:
- Before this tweak, we'd issue a second request for all fields.
  1. A request for `views,visitors` 
  2. A request for `views,visitors,likes,comments,post_titles`
- After tweak, we issue a second request for only the fields we need.
  1. A request for `views,visitors` 
  2. A request for `likes,comments,post_titles`

This optimization should alleviate the load on the API.

sgomes commented on client/my-sites/stats/stats-chart-tabs/index.jsx:
```diff
@@ -163,141 +91,87 @@ class StatModuleChartTabs extends Component {
 		areStatePropsEqual: compareProps( {
-			deep: [ 'activeTab', 'fullQuery', 'quickQuery' ],
+			shallow: [ 'activeTab', 'isActiveTabLoading' ],
+			deep: [ 'query', 'loadingTabs' ],
```
I think in general the preferred approach tends to be memoization on the selectors (using e.g. `createSelector`), rather than performing deep comparisons. Here it shouldn't be too bad, though, since the objects are pretty much known quantities and not too complex.

In any case, I'll take a look to see if any re-renders are happening and whether we need to make any changes here.

sgomes replied:
I did some checking, and this component does seem to be causing unnecessary re-renders. `chartData`, specifically, is causing re-renders further down the line, in the `Chart` and `ChartBarContainer` components, because the generated data isn't equal by reference even though it's equal by value.

In any case, feel free to leave this for me to look at after your PR, since the problem was already there :)

jsnmoon replied:
Good catch; I'll leave it alone for this PR. 

Feel free to mark me down as a reviewer for your next pass :)

sgomes commented:
Thanks for the testing instructions, @jsnmoon ! Are we specifically looking for calls to the `stats` endpoint here? If so, I'm only seeing one connection here, not two.

I'll come back to this PR on Monday with some performance analysis in terms of before/after. Hopefully We'll be able to quantize things a bit :)

jsnmoon commented:
> Thanks for the testing instructions, @jsnmoon ! Are we specifically looking for calls to the `stats` endpoint here? If so, I'm only seeing one connection here, not two.

@sgomes: There ought to be two requests to `/stats/visits`, like so:

<img width="1284" alt="Network requests from current production" src="https://user-images.githubusercontent.com/4044428/47585008-65323880-d919-11e8-8328-f35d50d440a5.png">

I updated the original post to clarify. Hope this helps!

sgomes commented:
Hi @jsnmoon !

I ran a bunch of tests today, and I couldn't find any evidence of performance improvements or any change in terms of number of calls to `/stats/visits`. I rolled back to 3dab7389a87dc574f472521fdfd5d8697402542d and at that commit there seem to be only two `visits` requests made as well, and with the same query string:

![image](https://user-images.githubusercontent.com/409615/47650374-f56db900-db77-11e8-8179-a0c50aeb316c.png)

I measured the performance using `perfmon` under a simulated 3G connection when loading the stats page from scratch (that is, navigating to it), when opening it for the first time (having navigated to a different page first), and when making repeated visits to it, and I couldn't find a meaningful performance delta in any of those scenarios.

Did I miss anything or do anything wrong? Is there perhaps a situation in which the benefits are more evident?

jsnmoon commented:
Hi @sgomes,

It's strange that the second request in your screenshot includes `visitors` in the query string. Is your setup running the latest commit (94a12bed4e787817e106ead692b5d2b61c3def4d)? 

I was able to reproduce the different queries in [this Calypso live branch](https://hash-94a12bed4e787817e106ead692b5d2b61c3def4d.calypso.live/). (For easy reference, here's [master running on a live branch](https://hash-564fd7b47a3fdc82166f7ac861b47f1b59ae2a37.calypso.live/).

<img width="674" alt="screen shot 2018-10-29 at 10 42 22 am" src="https://user-images.githubusercontent.com/4044428/47665719-6193f100-db67-11e8-8fdb-92cb54381964.png">

Note that `views,visitors` has been excluded from the second request in the screenshot above.

jsnmoon commented:
I ran some performance profiling at an internal P2 using Calypso.live at two hashes: 94a12bed4e787817e106ead692b5d2b61c3def4d and 3dab7389a87dc574f472521fdfd5d8697402542d. 

---

# Day (Avg of 5 requests)
## Without change
- Visitor request: 207.4 ms
- Supplemental request: 6.428s

## With change
- Visitor request: 170.2ms
- Supplemental request: 6.01s

## Delta
- Visitor request: -37.2ms
- Supplemental request: -418ms

# Week  (Avg of 10 requests)
## Without change
- Visitor request: 243.0ms
- Supplemental request: 7.54s

## With change
- Visitor request: 261.5ms
- Supplemental request: 7.91s

## Delta
- Visitor request: +18.5ms 
- Supplemental request: +370ms

# Month  (Avg of 10 requests)
## Without change
- Visitor request: 596.8ms
- Supplemental request: 9.597s

## With change
- Visitor request: 603.1ms
- Supplemental request: 8.93s

## Delta
- Visitor request: +6.3ms
- Supplemental request: -667ms

---

Takeaway: There might be an improvement in network performance, but it'd be hasty to make this conclusion given that there's a lot of variance between each network request (not to mention Week requests actually did worse). However, I believe we should push out this change since it removes [requesting redundant data fields](https://github.com/Automattic/wp-calypso/pull/27997/files#diff-b61543a6dc80031e3a9fcb3d8e280f5cR20) from the `stats/visits` API route.

sgomes commented:
Thanks, @jsnmoon ! That's great.

I agree that even if the performance numbers are inconclusive this PR should be merged, as the refactoring merits are sufficient on their own. 👍
